Implementation of LandTask

mbilker opened the pull request:
This is almost an exact copy of the Takeoff task except the conditional is switched.

 - I didn't see a delay before starting to descend so there is no check against `elapsed`.
 - The code could be simpler by setting the descending velocity initially and only having the conditional handle the task completion.

mbilker committed:
Implementation of LandTask

mbilker commented:
How do I initialize the state of the drone to an arbitrary height?

The controller has access to `self.state_controller` but the task's update method passes the `environment` variable to get the drone state using `state_controller.query('DroneState', environment)`.

hgarrereyn commented:
You could set an arbitrary height by modifying the drone's attributes directly: `environment.agent.z_pos = 1` for example.

Alternatively, you could test the TakeoffTask followed by the LandTask using the callback method.

hgarrereyn commented on roombasim/pittras/ai/land_demo_controller.py:
```diff
@@ -0,0 +1,12 @@
+        # PITTRAS_LAND_HEIGHT_TOLERANCE
+
+        self.task_controller.switch_task(
+            'LandTask'
```
make sure to export the LandTask in `roombasim/pittras/tasks/__init__.py` and add it to the `roombasim.pittras.config` file

hgarrereyn commented on roombasim/pittras/task/land_task.py:
```diff
@@ -0,0 +1,40 @@
+
+    def __init__(self):
+        # state machine
+        self._state = TakeoffTaskState.init
```
should be LandTaskState

hgarrereyn commented on roombasim/pittras/task/land_task.py:
```diff
@@ -0,0 +1,40 @@
+            self.complete(Task.SUCCESS)
+            return
+
+        environment.agent.control(self.zero_xy_vel, 0, control_z_vel)
```
self.zero_xy_vel doesn't exist, you can just use `[0, 0]`

hgarrereyn commented on roombasim/pittras/ai/land_demo_controller.py:
```diff
@@ -0,0 +1,12 @@
+
+from roombasim.ai import Controller
+
+class LandDemoController(Controller):
```
please export LandDemoController in `roombasim/pittras/ai/__init__.py`

hgarrereyn commented on roombasim/pittras/task/land_task.py:
```diff
@@ -0,0 +1,40 @@
+            # Stop descending if below the target height tolerance
+            self._state = LandTaskState.done
+
+            self.complete(Task.SUCCESS)
```
you need to import TaskState (from `roombasim.ai`) and use `TaskState.SUCCESS`

mbilker commented:
Yep, I didn't do a thorough review nor testing, I just wanted an answer to set the initial height. Will fix.

mbilker committed:
LandTask: fix land task demo controller

amiller27 commented on roombasim/pittras/config.py:
```diff
@@ -60,6 +61,12 @@
+PITTRAS_LAND_VELOCITY = -0.2
+
+# height below which landing is considered finished
+PITTRAS_LAND_HEIGHT_TOLERANCE = 0.15
```
This parameter isn't actually used on the drone, RoombaSIM should probably just check if the drone is going to move into or below the floor and check if the landing is complete that way.  Then when the landing is complete it should probably set the height to 0 exactly.

mbilker replied:
Hmm, should I check if the `height` is equal to zero in the task to determine if the task has completed?

amiller27 replied:
Yes, that will work.  The current implementation in `agent/drone.py` sets the height to exactly `0` when you hit the floor, but I'm not sure if you want to make that the defined behavior or not.  That's a question for @hgarrereyn

amiller27 commented on roombasim/pittras/task/land_task.py:
```diff
@@ -0,0 +1,36 @@
+
+from roombasim.ai import Task, TaskState
+
+class LandTaskState:
```
This should inherit from `object`, no?

mbilker replied:
I think Python 3 heavily discouraged a class from inheriting from nothing so it should inherit from `object`.

amiller27 replied:
In Python 2 you should really inherit from `object`, in Python 3 you don't have to but it doesn't hurt if you do

zacyu replied:
Have we agreed on which python version we are primarily supporting? If we still want compatibility for both python 2 and 3, I think it will be better to have the inheritance explicit.

amiller27 replied:
We _need_ support for python 2, python 3 support is up to you guys

mbilker committed:
LandTask: inherit LandTaskState from object

mbilker committed:
LandTask: remove landing height tolerance

- not used in the motion planner
- for the simulator, zero height is declared as landed

mbilker commented:
PTAL :smile:

hgarrereyn commented on roombasim/pittras/ai/land_demo_controller.py:
```diff
@@ -0,0 +1,36 @@
+'''
+
+from roombasim.ai import Controller
+from roombasim.ai.task import TaskState
```
TaskState is exported by roombasim.ai so you can do:
`from roombasim.ai import Controller, TaskState`

hgarrereyn commented on roombasim/pittras/ai/land_demo_controller.py:
```diff
@@ -0,0 +1,36 @@
+
+        # TODO(mbilker): is there a way to initialize the drone to an
+        # arbitrary height above PITTRAS_LAND_HEIGHT_TOLERANCE instead of using
+        # TakeoffTask to get the height up
```
The following one-liner should do the trick:

```py
(lambda cfg:(setattr(cfg,'AGENT',(lambda agent:(lambda *p:(lambda ref:(lambda _,x: x)(setattr(ref,'z_pos',1.5),ref))(agent(*p))))(getattr(cfg,'AGENT')))))(__import__('roombasim.config').config)
```

In all seriousness, I think I'll add some functionality to specify where to spawn the drone via the CLI.

mbilker replied:
For the sake of testing, would it make more sense to have a random initialization in the controller that can be over-ridden by the CLI?

To do this in the controller, how do I access `environment` if that is even possible in the current implementation of RoombaSIM?

amiller27 replied:
If you're looking for a hack for testing purposes, you can just set `environment.agent.z_pos` to whatever you want the altitude to be

mbilker replied:
Should I do that with the `LandTaskState.init` state in `LandTask`?

amiller27 replied:
You could do that, yes.  We really don't want that to be a permanent solution, probably don't want to merge code using that, but for testing it'll work

mbilker replied:
I really would like a way to do this in the controller's `setup` method.

amiller27 replied:
It's not something for the controller to set

amiller27 replied:
Nvm, forgot what the controller is

mbilker replied:
True, I could just use the `XYZTranslationTask` to navigate to the spot for setup and test the `LandTask` from there.

mbilker replied:
Oh, I think its appropriate for the controller to set the location for startup, especially since the controller is what is specified by the command line and could be set to a testing controller.

amiller27 replied:
Yeah I forgot what RoombaSIM calls a "controller."  I still don't think it's a great idea to let the controllers set the position directly, it's already a parameter for the agent on startup we would just have to add a parameter in the cli

amiller27 replied:
Added the option to specify on command line in the `add-location-option` branch, you can check it out and cherry-pick or rebase or merge it or whatever

hgarrereyn replied:
Nice, this looks like a good way to do it. @mbilker can you merge `add-location-option` into this branch?

hgarrereyn replied:
This comment seems a bit buried, I'll go ahead and merge this

hgarrereyn commented on roombasim/pittras/task/land_task.py:
```diff
@@ -0,0 +1,36 @@
+        height = drone_state['z_pos']
+
+        # Check if we reached the target landing height tolerance
+        if height > 0:
```
this is fine for now since position will be set to zero if the drone goes below the ground. However, in the future we might implement some leg sensors that could give us a yes/no reading for if we are on the ground

mbilker replied:
Would this be blocking on this change or should this be added to some planned changes file going forward?

hgarrereyn replied:
fine for now

hgarrereyn reviewed (commented):
Looks good and it seems to work if the drone is set to a position off the ground.

mbilker committed:
LandTask: roombasim.ai imports TaskState

amiller27 committed:
Added option to specify start location for drone

mbilker committed:
LandTask: add note regarding drone landing completion detection

mbilker committed:
LandTask: controller only executes LandTask

- add note regarding potential to initialize the drone environment to an arbitrary height for testing

hgarrereyn committed:
Merge branch 'add-location-option' into land_task
- needed for land_task testing

hgarrereyn committed:
Merge branch 'master' into land_task
- fixed LandDemoController callback

mbilker commented:
Yes, those comments got buried. I could only see them email.
